nutils/evaluable.py

```diff
@@ -1774,41 +1776,36 @@ def _intbounds_impl(self):
 class Add(Array):
 
     def __init__(self, funcs: types.frozenmultiset):
-        assert isinstance(funcs, types.frozenmultiset) and len(funcs) == 2, f'funcs={funcs!r}'
-        self.funcs = funcs
-        func1, func2 = funcs
-        assert equalshape(func1.shape, func2.shape) and func1.dtype == func2.dtype != bool, 'Add({}, {})'.format(func1, func2)
-        super().__init__(args=tuple(self.funcs), shape=func1.shape, dtype=func1.dtype)
+        assert isinstance(funcs, types.frozenmultiset), f'funcs={funcs!r}'
+        func, *others = funcs
+        assert all(equalshape(f.shape, func.shape) and f.dtype == func.dtype for f in others)
+        self.funcs = tuple(funcs)
+        super().__init__(args=tuple(funcs), shape=func.shape, dtype=func.dtype)
 
     @cached_property
     def _inflations(self):
-        func1, func2 = self.funcs
-        func2_inflations = dict(func2._inflations)
-        inflations = []
-        for axis, parts1 in func1._inflations:
-            if axis not in func2_inflations:
-                continue
-            parts2 = func2_inflations[axis]
-            dofmaps = set(parts1) | set(parts2)
-            if (len(parts1) < len(dofmaps) and len(parts2) < len(dofmaps)  # neither set is a subset of the other; total may be dense
-                    and self.shape[axis].isconstant and all(dofmap.isconstant for dofmap in dofmaps)):
-                mask = numpy.zeros(int(self.shape[axis]), dtype=bool)
-                for dofmap in dofmaps:
-                    mask[dofmap.eval()] = True
-                if mask.all():  # axis adds up to dense
-                    continue
-            inflations.append((axis, types.frozendict((dofmap, util.sum(parts[dofmap] for parts in (parts1, parts2) if dofmap in parts)) for dofmap in dofmaps)))
-        return tuple(inflations)
+        inflations_per_axis = [[] for i in range(self.ndim)]
+        for func in self.funcs:
+            for axis, parts in func._inflations:
+                inflations_per_axis[axis].append(parts)
+        return tuple((axis,
+            types.frozendict((dofmap, add(*(parts[dofmap] for parts in parts_per_func if dofmap in parts)))
+                for dofmap in functools.reduce(set.union, map(set, parts_per_func))))
+                    for axis, parts_per_func in enumerate(inflations_per_axis)
+                        if len(parts_per_func) == len(self.funcs))
```

Comment:
The indentation really confuses me. The last two lines belong to the generator starting at line 1791, but the line before, with less indentation, belongs to the inner generator starting at line 1791. Also it is hard to see where the tuple `(axis, ...)`, the item of the outer generator, ends. Can you please reconsider this?

Reply:
I reformatted it to make more sense but at the expense of a very long line.